witness/src/Dscp/Snowdrop/Storage/Avlp.hs

```diff
@@ -0,0 +1,311 @@
+    }
+
+-- | This is a plugin to run actions in RocksDB.
+rocksDBPlugin :: RDB.RocksDB -> DBPlugin (ReaderT RDB.RocksDB)
+rocksDBPlugin rdb = DBPlugin (`runReaderT` rdb)
+
+-- | This is a plugin to run actions in RocksDB.
+-- pureDBPlugin :: AVL.Params h k v => DBPlugin (Pure.Store h k v)
+-- pureDBPlugin = DBPlugin Pure.run
+
+-- | Following constraints contain 't', because we only use
+--   AVL actions in 't m' contexts.
+type AVLMutates h k v t m =
+    ( CanSerialise h k v
+    , AVL.Mutates  h k v (t m)
+    , MonadIO    m
+    , MonadCatch m
+    )
+
+-- | Version of 'AVLMutates' to use outside of this module.
+type AVLMutates' h k v m =
+    ( CanSerialise h k v
+    , AVL.Mutates  h k v m
+    )
+
+-- | Basic requirements to implement RocksDB storage.
+type AVLBase h k v m =
+    ( CanSerialise h k v
+    , MonadIO m
+    , AVL.Base h k v m
+    )
+
+-- | So we can put (hash -> node) relation into RocksDB.
+type CanSerialise h k v =
+    ( Serialise h
+    , Serialise (AVL.MapLayer h k v h)
+    )
+
+-- | Instance for reading from the database.
+instance
+    ( AVLBase h k v m
+    , MonadReader ctx m
+    , HasLens' ctx RDB.RocksDB
+    )
+      =>
+    AVL.KVRetrieve h (AVL.MapLayer h k v h) m
+  where
+    retrieve h = do
+        RDB.get h >>= maybe
+            (throwM $ AVL.NotFound h)
+            return
+
+-- | Instance for reading from the database.
+instance
+    ( AVLBase h k v m
+    , MonadReader ctx m
+    , HasLens' ctx RDB.RocksDB
+    )
+      =>
+    AVL.KVStore h (AVL.MapLayer h k v h) m
+  where
+    massStore = RDB.put
+
+-- | Instance for getting/setting a current root in the database.
+-- | Also, for overwriting, if needed.
+instance
+    ( AVLBase h k v m
+    , MonadReader ctx m
+    , HasLens' ctx RDB.RocksDB
+    )
+      =>
+    AVL.KVMutate h (AVL.MapLayer h k v h) m
+  where
+    getRoot = RDB.get (rootName @h @k @v)
+        >>= maybe (throwM AVL.NoRootExists) return
+
+    setRoot newRoot = do
+        RDB.put [(rootName @h @k @v, newRoot)]
+
+    erase h = do
+        RDB.delete [h]
+
+-- | A tip name for AVL tree with given argument types.
+rootName :: forall h k v . (Typeable h, Typeable k, Typeable v) => ByteString
+rootName
+    = "tree root of Map ("
+    <> show (typeOf (error "shouldn't fire" :: h))
+    <> ") ("
+    <> show (typeOf (error "shouldn't fire" :: k))
+    <> ") ("
+    <> show (typeOf (error "shouldn't fire" :: v))
+    <> ")"
+
+-- | Maybe is there due to SD imposing Default constrains
+--   on accum /because reasons/.
+type AVLChgAccum h k v = Maybe (RealAVLAccum h k v)
+
+data RealAVLAccum h k v =
+  RealAVLAccum
+    { acaMap     :: AVL.Map h k v
+    -- ^ AVL map, which contains avl tree with most-recent updates
+    , acaTouched :: Set AVL.Revision
+    -- ^ Set of nodes, which were touched during all of change operations applied on tree
+    }
+
+-- | This function is written out of inability to to proper coding with SD.
```

Comment:
choo choo. Also the comment doesn't really make a lot of sense. "This function is written because <problem in SD which is not properly described>", but what does it do/where is it used?